### PR TITLE
Added Watecco Triphaso

### PR DESCRIPTION
**Pull request by J2thatsme**

### CODE REVIEW

File: types/watecco/triphaso/uplink.js [hunk 1]
```diff
@@ -0,0 +1,1124 @@
...
+          decoded.zclheader.max.unity = "minutes";
+        } else {
+          decoded.zclheader.max.value = bytes[9] * 256 + bytes[10];
+          decoded.zclheader.max.unity = "seconds";
+        }
+      }
+    }
+  }
+
+  // emit("sample", { topic: "default", data: decoded.data });
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:1123]: add or remove

File: types/watecco/triphaso/uplink.js [hunk 2]
```diff
@@ -0,0 +1,1124 @@
+const ST_UNDEF = 0;
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:1]: add a link to the source decoder

File: types/watecco/triphaso/uplink.js [hunk 3]
```diff
@@ -0,0 +1,1124 @@
...
+  }
+  return lastTimestamp;
+}
+
+/**
+ * Translate brUncompress output data to expected structure
+ */
+function adaptToExpectedFormat(out, argList, batchAbsoluteTimestamp) {
+  const returnedGlobalObject = {
+    // batch_counter: out.batch_counter,
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:678]: review this commented code

File: types/watecco/triphaso/uplink.js [hunk 4]
```diff
@@ -0,0 +1,1124 @@
...
+
+/**
+ * Compute data absolute timestamp from batch absolute timestamp (bat), batch
+ * relative timestamp (brt) and data relative timestamp (drt)
+ */
+function computeDataAbsoluteTimestamp(bat, brt, drt) {
+  return new Date(new Date(bat) - (brt - drt) * 1000).toISOString();
+}
+
+try {
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:722]: remove this as nothing needs to be exported

File: types/watecco/triphaso/uplink.js [hunk 5]
```diff
@@ -0,0 +1,1124 @@
...
+      decoded.zclheader.endpoint =
+        ((bytes[0] & 0xe0) >> 5) | ((bytes[0] & 0x06) << 2);
+      // command ID
+      cmdID = bytes[1];
+      decoded.zclheader.cmdID = decimalToHex(cmdID, 2);
+      // Cluster ID
+      clusterdID = bytes[2] * 256 + bytes[3];
+      decoded.zclheader.clusterdID = decimalToHex(clusterdID, 4);
+
+      // decode report and read atrtribut response
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:826]: typo attribute

File: types/watecco/triphaso/uplink.js [hunk 6]
```diff
@@ -0,0 +1,1124 @@
...
+      if ((cmdID === 0x0a) | (cmdID === 0x8a) | (cmdID === 0x01)) {
+        const stdData = {};
+        const tab = {};
+
+        // Attribut ID
+        attributID = bytes[4] * 256 + bytes[5];
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:832]: typo attribute

File: types/watecco/triphaso/uplink.js [hunk 7]
```diff
@@ -0,0 +1,1124 @@
...
+            UintToInt(bytes[index + 13] * 256 + bytes[index + 14], 2) / 10;
+          variables.irmsC =
+            UintToInt(bytes[index + 15] * 256 + bytes[index + 16], 2) / 10;
+          variables.angleC =
+            UintToInt(bytes[index + 17] * 256 + bytes[index + 18], 2) / 10;
+
+          emit("sample", { topic: "variables", data: variables });
+        }
+
+        // decoded.data = tab;
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:1076]: review the commented code

**Review by mariusgiger — CHANGES_REQUESTED** [review]
looks good, but please review the commented sections once more, also there is a case where no output is emitted

File: types/watecco/triphaso/phase_a.schema.json [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      "type": "number"
+    },
+    "pReactiveEnergyA": {
+      "title": "Positive reactive energy A",
+      "unit": "var",
+      "description": "Positive reactive energy A",
+      "type": "number"
+    },
+    "nReactiveEnergyA": {
+      "title": "Negativee reactive energy A",
```
Comment (mariusgiger) [inline types/watecco/triphaso/phase_a.schema.json:28]: change all instances of negativee to negative

File: types/watecco/triphaso/uplink.js [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+        } else {
+          decoded.zclheader.alarm = 0;
+        }
+
+        let index = 0;
+        // data index start
+        if ((cmdID === 0x0a) | (cmdID === 0x8a)) {
+          index = 7;
+        }
+        // if (cmdID === 0x01) {index = 8; decoded.zclheader.status = bytes[6];}
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:165]: add or remove this

File: types/watecco/triphaso/uplink.js [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+      tempHexStr = `0${tempHexStr}`;
+    }
+    lora.payload += tempHexStr;
+  }
+
+  if (port === 125) {
+    // batch
+    decodedBatch = !(bytes[0] & 0x01);
+
+    // trame standard
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:106]: standard frame?

File: types/watecco/triphaso/uplink.js [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+          phase[`nReactivePowerVar${key}`] = UintToInt(
+            bytes[index + 29] * 256 * 256 * 256 +
+              bytes[index + 30] * 256 * 256 +
+              bytes[index + 31] * 256 +
+              bytes[index + 32],
+            4,
+          );
+
+          emit("sample", {
+            topic: `phase_${topic}`,
```
Comment (mariusgiger) [inline types/watecco/triphaso/uplink.js:240]: use more speaking variable names: e.g. phase to describe the phase as a string and phaseData to describe the resulting payload

File: types/watecco/triphaso/variables.schema.json [hunk 12]
```diff
@@ -0,0 +1,75 @@
+{
+  "$id": "https://akenza.io/watecco/triphaso/variables.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Phase variables",
+  "processingType": "uplink_decoder",
+  "topic": "variables",
```
Comment (mariusgiger) [inline types/watecco/triphaso/variables.schema.json:6]: what's the difference to the topic phase_variables?

**Review by mariusgiger — COMMENTED** [review]
thanks for the update - please check this one again, there seem to be one or two things to follow up

File: types/watecco/triphaso/variables.schema.json (hunk 12 above)
Comment (J2thatsme) [reply types/watecco/triphaso/variables.schema.json]: I splitted them up

**Comment by J2thatsme** [comment]
Im going to create a new pr for this. Some clients already use a custom devicetype for this device.
In that, we have our new features already implemented.

**Review by mariusgiger — APPROVED** [review]

**Pull request merged**